graphql_compiler/tests/test_data_tools/data_tool.py

```diff
@@ -16,3 +37,60 @@ def generate_data(client):
                     continue
 
                 client.command(sanitized_command)
+
+
+@retry(tries=20, timeout=1)  # pylint: disable=no-value-for-parameter
+def init_sql_integration_test_backends():
+    """Connect to and open transaction on each SQL DB under test."""
+    sql_test_backends = {}
+    for backend_name, connection_string in six.iteritems(SQL_BACKEND_TO_CONNECTION_STRING):
+        engine = create_engine(connection_string)
+        # MYSQL and MARIADB do not have a default DB so a DB must be created
+        if backend_name in {TestBackend.MYSQL, TestBackend.MARIADB}:
+            # safely create the DB
+            engine.execute(text('DROP DATABASE IF EXISTS animals; CREATE DATABASE animals'))
+            # update the connection string and engine to connect to this DB specifically
+            connection_string = connection_string + u'/animals'
+            engine = create_engine(connection_string)
+        connection = engine.connect()
+        transaction = connection.begin()
```

Comment:
This seems a bit strange, starting a transaction and passing it around like that. I think it'd be better practice if the component that starts the transaction also makes sure that it ends, either with a commit or an abort. It's also probably a good idea to have isolation between tests by running each test in a separate transaction.

Reply:
Good call, I'll leave creating engines and connections in the module scope, and start transactions at the function scope.

Reply:
I forgot here, the outer transaction was so that after the integration test is over, the data is cleared by rolling back the transaction. Instead there is now a nested transaction for each test function that is run, which is rolled back after the test.

As far as what you mean by passing around a transaction, this is managed by the test fixture that calls this, and that fixture is also responsible for calling rollback. Should it be more explicit than that? That would be the `sql_integration_data` function, which calls `init` here and then `teardown`, which then begin and rollback the transactions, respectively.

Reply:
> I forgot here, the outer transaction was so that after the integration test is over, the data is cleared by rolling back the transaction. Instead there is now a nested transaction for each test function that is run, which is rolled back after the test.

I'm not sure I understand this. Is the intention to run each test in its own nested transaction, within the single transaction that wraps all tests? What's the benefit of that versus not nesting transactions and relying on the fixture to clean up after itself anyway? I am concerned about both the complexity as well as potential cross-test issues here, such as the database deciding that the outer transaction must abort (for any reason, as it is allowed to do so) and preventing all remaining tests from running at all.

As for what becomes part of the `SqlTestBackend`, I wasn't clear on why we're sending the connection string, and also the engine created from said connection string, as well as a connection from said engine. Shouldn't just one of them be enough, depending on how much of the test infrastructure we want tests to share?

Happy to chat over video call if that would be faster and easier here.